lib/os/memcpy32.c

```diff
@@ -0,0 +1,49 @@
+#include <sys/util.h>
+#include <sys/__assert.h>
+#include <stdint.h>
+#include <toolchain.h>
+
+/*
+ * Copyright (c) 2021 Pete Dietl
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define IS_ALIGNED(X) (!((uintptr_t)X & (sizeof(uint32_t) - 1)))
+
+void *memcpy32(void *restrict dst0, const void *restrict src0, size_t len)
+{
+	__ASSERT(len % 4 == 0, "len must be a multiple of 4!");
+
+	typedef uint32_t __may_alias u32_alias;
```

Comment:
Scary gcc attribute. But I assume all supported toolchains are in CI.

Reply:
Yes this is scary, but there is no other way to get around strict aliasing that I know of :(

Reply:
Should we add `ZRESTRICT` to match `memcpy()` in `lib/libc/minimal/include/string.h`?

Reply:
Done

Reply:
> Scary gcc attribute. But I assume all supported toolchains are in CI.

actually they are not.

Only the Zephyr SDK and clang is in CI, but none of the other commercial toolchains.

Other toolchains are gcc or clang based, so if this works with those two, I expect other toolchain should work.
@pdietl can you ensure that this will be tested by CI for minimum Zephyr SDK and clang.